File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        ListNode prev = head;
        ListNode cur = head.next;
        ListNode next = head.next.next;

        int firstCritical=-1 , lastCritical=-1;
        int index=1;
        int minDist=Integer.MAX_VALUE;

        while(next!=null){
             
             if(cur.val > prev.val && cur.val > next.val 
                || cur.val < prev.val && cur.val < next.val){
                
                if(firstCritical== -1){
                    firstCritical = index;
                }else{ //index=2
                     minDist = Math.min(minDist , Math.abs(index-lastCritical));
                }
                lastCritical=index;
             }
             prev = cur;
             cur = next;
             next = next.next;
             index++;
        }

        if(lastCritical==firstCritical){
            return new int [] {-1, -1};
        }

        int maxDist= Math.abs(firstCritical-lastCritical);

        return new int[] {minDist , maxDist};
    }

}

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {
        ListNode* temp = head;
        vector<int> arr;

        while (temp) {
            arr.push_back(temp->val);
            temp = temp->next;
        }

        vector<int> critical;
        for (int i = 1; i < arr.size() - 1; i++) {
            if ( arr[i] > arr[i-1] && arr[i] > arr[i + 1] ||
                arr[i] < arr[i-1] && arr[i] < arr[i + 1]) {
                critical.push_back(i);
            }
        }

        if (critical.size() < 2) {
            return {-1, -1};
        }

        int minDist = 1e9;

        for (int i = 1; i < critical.size(); i++) {
            minDist = min(minDist, critical[i] - critical[i - 1]);
        }

        int maxDist = critical.back() - critical.front();

        return {minDist, maxDist};
    }
};